### PR TITLE
feat: Partial load stations in dropdown with option to load them all

### PR DESCRIPTION
**Pull request by SirBerg**
This changes the dropdown on the station select page to only show the first 800 stations and only load every other station once the user clicks a button.

### CODE REVIEW

File: src/components/stationsSelect.tsx [hunk 1]
```diff
@@ -36,6 +38,87 @@ export function StationSelect(props: Props) {
...
+
+  //runs everytime the dropdown is opened
+  React.useEffect(() => {
+    if(!availableStations) return
+    setAllStationsLoaded(false)
+    //setting the dropdown to the first 800 stations to achieve a faster opening of the dropdown
+    setDropDownCurrentState(
+      <CommandGroup>
+        {/*the .filter function here limits the station array to the first 800*/}
+        {availableStations.filter((item, idx) => idx < 800).map((station) => (
```
Comment (michidk) [inline src/components/stationsSelect.tsx:90]: its more performant to use .slice instead of .filter for that

File: src/lib/stations.ts [hunk 2]
```diff
@@ -30,4 +30,4 @@ export async function pollStations() {
       id: station.id,
     };
   });
-}
+}
```
Comment (michidk) [inline src/lib/stations.ts:33]: this file wasn't changed

File: src/components/stationsSelect.tsx [hunk 3]
```diff
@@ -48,61 +131,46 @@ export function StationSelect(props: Props) {
...
+                    {dropDownCurrentState}
+                  </ScrollArea>
+                </Command>
+              </PopoverContent>
+            </Popover>
+          </>
+        )}
+      </>
   );
-}
```
Comment (michidk) [inline src/components/stationsSelect.tsx:108]: please make sure to commit with `LF` and not `CR FL`

File: src/components/stationsSelect.tsx [hunk 4]
```diff
@@ -48,61 +131,46 @@ export function StationSelect(props: Props) {
...
+                  {boxLabel}
+                  <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
+                </Button>
+              </PopoverTrigger>
+              <PopoverContent className="w-[200px] p-0">
+                <Command>
+                  <CommandInput placeholder="Search station..." />
+                  <ScrollArea className="h-96">
+                    {
+                      allStationsLoaded ? <></> :
```
Comment (michidk) [inline src/components/stationsSelect.tsx:158]:
```suggestion
                      allStationsLoaded ? <></> :
```

File: src/components/stationsSelect.tsx [hunk 5]
```diff
@@ -48,61 +131,46 @@ export function StationSelect(props: Props) {
...
-                </ScrollArea>
-              </Command>
-            </PopoverContent>
-          </Popover>
-        </>
-      )}
-    </>
+      {
+        //making sure that availableStations is always available at this point
+        !availableStations || availableStations.length == 0 ?
```
Comment (michidk) [inline src/components/stationsSelect.tsx:136]:
```suggestion
        !availableStations || availableStations.length == 0 ?
```

File: src/components/stationsSelect.tsx (hunk 1 above)
Comment (michidk) [reply src/components/stationsSelect.tsx:90]: also can we make the 800 a constant?

File: src/components/stationsSelect.tsx [hunk 6]
```diff
@@ -36,6 +38,87 @@ export function StationSelect(props: Props) {
...
+      </CommandGroup>
+    )
+  }
+
+  //runs everytime the dropdown is opened
+  React.useEffect(() => {
+    if(!availableStations) return
+    setAllStationsLoaded(false)
+    //setting the dropdown to the first 800 stations to achieve a faster opening of the dropdown
+    setDropDownCurrentState(
```
Comment (michidk) [inline src/components/stationsSelect.tsx:87]: I dont think storing that much html in the state is a good idea. I would just populate `availableStations` with only 800 entries from the start (or introduce a new variable and call that `availableStationsTrimmed`. then once the button is clicked, we change the list to be all stations

**Review by michidk — CHANGES_REQUESTED** [review]

File: src/components/stationsSelect.tsx (hunk 1 above)
Comment (SirBerg) [reply src/components/stationsSelect.tsx:90]: but .slice removes everything that comes after the 800 stations so we would have to query the API again. .filter doesn't do that. And yes I can make the 800 a constant :)

Comment (michidk) [reply src/components/stationsSelect.tsx:90]: No `availableStations.slice(0, 800).map(...` is the exact same as `availableStations.slice(0, 800).map(...` but less performant

File: src/lib/stations.ts (hunk 2 above)
Comment (michidk) [reply src/lib/stations.ts:33]:
```suggestion
}
```

File: src/components/stationsSelect.tsx (hunk 1 above)
Comment (SirBerg) [reply src/components/stationsSelect.tsx:90]: Oh I read `.splice` instead of `.slice` my bad

File: src/components/stationsSelect.tsx (hunk 6 above)
Comment (SirBerg) [reply src/components/stationsSelect.tsx]: Is the updated version OK for you?

File: .gitattributes [hunk 7]
```diff
@@ -0,0 +1 @@
+text=auto
```
Comment (michidk) [inline .gitattributes:1]: Recommend to set the following in your gitconfig:
```
[core]
    autocrlf = input
    eol = lf
```

File: .gitignore [hunk 8]
```diff
@@ -5,6 +5,7 @@
 /.pnp
 .pnp.js
 .yarn/install-state.gz
+package-lock.json
```
Comment (michidk) [inline .gitignore:8]: please don't do this

File: .gitignore [hunk 9]
```diff
@@ -36,3 +37,4 @@ yarn-error.log*
 next-env.d.ts
 
 .cache
+.idea
```
Comment (michidk) [inline .gitignore:40]: cringe bro

File: src/app/page.tsx [hunk 10]
```diff
@@ -1,11 +1,11 @@
 "use server";
 
-import { getStations } from "@/lib/stations";
+import { getStations, Station } from "@/lib/stations";
 import { StationUrl } from "@/components/stationUrl";
 
 export default async function Page() {
   try {
-    const stations = await getStations();
+    const stations:Station[] = await getStations();
```
Comment (michidk) [inline src/app/page.tsx:8]:
```suggestion
    const stations: Station[] = await getStations();
```

Comment (michidk) [inline src/app/page.tsx:8]:
```suggestion
    const stations: Array<Station> = await getStations();
```

File: src/components/stationsSelect.tsx [hunk 11]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
 import { Station } from "@/lib/stations";
 
 type Props = {
-  availableStations: Station[];
+  availableStations: Station[] | null;
```
Comment (michidk) [inline src/components/stationsSelect.tsx:26]: should not be optional

File: src/components/stationsSelect.tsx [hunk 12]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
...
   selectedStations: Station[];
   onSelect: (station: Station) => void;
 };
 
-export function StationSelect(props: Props) {
+
+
+export function StationSelect(props: Props):React.JSX.Element {
```
Comment (michidk) [inline src/components/stationsSelect.tsx:33]: I think this type can be well infered

File: src/components/stationsSelect.tsx [hunk 13]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
...
     onSelect,
+  }: {
+    availableStations: Station[] | null;
+    selectedStations: Station[];
+    onSelect: (station: Station) => void;
   } = props;
 
-  const [open, setOpen] = useState(false);
+  const trimmedStations:Station[] = availableStations ? availableStations.slice(0, maxStationsInitialLoad) : []
+  const [open, setOpen]: [open: boolean, setOpen: React.Dispatch<React.SetStateAction<boolean>>] = useState(false);
```
Comment (michidk) [inline src/components/stationsSelect.tsx:46]: usually not typed

File: src/components/stationsSelect.tsx [hunk 14]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
...
+  const [dropDownCurrentState, setDropDownCurrentState] : [dropDownCurrentState: React.JSX.Element, setDropDownCurrentState: React.Dispatch<React.SetStateAction<React.JSX.Element>>] = useState(<></>)
+  const [allStationsLoaded, setAllStationsLoaded] : [allStationsLoaded: boolean, setAllStationsLoaded: React.Dispatch<boolean>] = useState(false)
+
+  function renderStations(stations:Station[]):React.JSX.Element{
```
Comment (michidk) [inline src/components/stationsSelect.tsx:50]:
```suggestion
  function renderStations(stations: Array<Station>) {
```

File: src/components/stationsSelect.tsx [hunk 15]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
...
+                />
+                {station.name}
+              </CommandItem>
+          ))}
+        </CommandGroup>
+    )
+  }
+
+  //runs everytime the dropdown is opened
+  React.useEffect(():void => {
```
Comment (michidk) [inline src/components/stationsSelect.tsx:86]:
```suggestion
  React.useEffect(() => {
```

File: src/lib/stations.ts [hunk 16]
```diff
@@ -1,33 +1,46 @@
...
+  products: string[],
+  latitude: number,
+  longitude: number,
+}
+
+
 let stations: Station[] = [];
 
-export async function getStations() {
+export async function getStations():Promise<Station[]> {
```
Comment (michidk) [inline src/lib/stations.ts:25]:
```suggestion
export async function getStations(): Promise<Array<Station>> {
```

File: src/lib/stations.ts [hunk 17]
```diff
@@ -1,33 +1,46 @@
...
 
-const API_TIMEOUT = 1000 * 5; // 5 seconds
+const API_TIMEOUT:number = 1000 * 5; // 5 seconds
 
 export type Station = {
   name: string;
   id: string;
 };
 
+export type apiReturnStation = {
```
Comment (michidk) [inline src/lib/stations.ts:10]: typically types are PascalCase

File: src/lib/stations.ts [hunk 18]
```diff
@@ -1,33 +1,46 @@
...
+export async function getStations():Promise<Station[]> {
   if (stations.length == 0) {
     stations = await pollStations();
   }
   return stations;
 }
 
-export async function pollStations() {
-  const response = await axios.get("https://www.mvg.de/.rest/zdm/stations", { timeout: API_TIMEOUT });
+export async function pollStations():Promise<Station[]>{
```
Comment (michidk) [inline src/lib/stations.ts:32]: arr

File: src/lib/stations.ts [hunk 19]
```diff
@@ -1,33 +1,46 @@
...
+  const response:AxiosResponse<apiReturnStation[]> = await axios.get("https://www.mvg.de/.rest/zdm/stations", { timeout: API_TIMEOUT });
   if (response.status !== 200) {
     throw new Error("Could not get stations");
   }
 
-  const stations: any[] = response.data;
+  const stations: apiReturnStation[] = response.data;
 
-  return stations.map((station: any) => {
+  return stations.map((station: apiReturnStation):Station => {
```
Comment (michidk) [inline src/lib/stations.ts:40]: spaces, please run prettier

File: src/components/stationsSelect.tsx [hunk 20]
```diff
@@ -23,19 +23,74 @@ import { cn } from "@/lib/utils";
...
+
+  function renderStations(stations:Station[]):React.JSX.Element{
+    return (
+        <CommandGroup>
+          {stations.map((station:Station) => (
+              <CommandItem
+                  key={station.id}
+                  value={station.name}
+                  onSelect={(currentValue:string):void => {
+                    const selectedStation:Station | undefined = availableStations ? availableStations.find(
```
Comment (michidk) [inline src/components/stationsSelect.tsx:58]: cant be undefined

File: src/components/stationsSelect.tsx [hunk 21]
```diff
@@ -48,61 +103,46 @@ export function StationSelect(props: Props) {
...
-                </ScrollArea>
-              </Command>
-            </PopoverContent>
-          </Popover>
-        </>
-      )}
-    </>
+      {
+        //making sure that availableStations is always available at this point
+        !availableStations || availableStations.length == 0 ?
```
Comment (michidk) [inline src/components/stationsSelect.tsx:108]: agh it always exists

**Review by michidk — CHANGES_REQUESTED** [review]

File: src/components/stationsSelect.tsx [hunk 22]
```diff
@@ -28,14 +28,29 @@ type Props = {
   onSelect: (station: Station) => void;
 };
 
-export function StationSelect(props: Props) {
+
+
```
Comment (michidk) [inline src/components/stationsSelect.tsx:32]:
```suggestion
```

File: src/components/stationsSelect.tsx [hunk 23]
```diff
@@ -28,14 +28,29 @@ type Props = {
...
+export function StationSelect(props: Props):React.JSX.Element {
```
Comment (michidk) [inline src/components/stationsSelect.tsx:33]:
```suggestion
export function StationSelect(props: Props): React.JSX.Element {
```

File: src/components/stationsSelect.tsx [hunk 24]
```diff
@@ -28,14 +28,29 @@ type Props = {
...
+  const maxStationsInitialLoad:number = 500
```
Comment (michidk) [inline src/components/stationsSelect.tsx:34]:
```suggestion
  const maxStationsInitialLoad: number = 500
```

**Review by michidk — CHANGES_REQUESTED** [review]
Can you run this through prettier once?

**Comment by michidk** [comment]
Very cool, thanks!

**Pull request merged**